README.md
# Big Dipper 2.0 ✨ (Cosmos Based Chains)

Big Dipper is an open-source block explorer and token management tool serving over 10 proof-of-stake blockchains. It has been forked more than 100 times on GitHub and has served audiences from 140 countries and regions.

This repo contains the **UI of Big Dipper 2.0** only. Read our official documentation at [http://docs.bigdipper.live/](http://docs.bigdipper.live/)

## Developer Guide

### Build and start using Docker

To build and run the application, you can use the [existing Docker Compose specification file](docker-compose.yml) or modify it to your needs:

```bash
docker compose up --detach
```

If you don't want to use Docker Compose for the build, you can also invoke this using just the Docker CLI:

```bash
docker build . --target runner -t bigdipper:local
```

To start the service, without doing a local build (for example, if you're already built an image previously), you can use:

```bash
docker compose up --detach --no-build
```

### Build and start using Yarn

This app can only be built and run using Yarn. Trying to use NPM doesn't work since dependency resolution works differently in NPM.

```bash
yarn install
yarn build
```

To start, you would need to provide environment variables. This would be easier to do in Docker instead.

```bash
yarn start
```

## Issue Reporting

For UI related issues please report it here [https://github.com/forbole/big-dipper-2.0-cosmos/issues](https://github.com/forbole/big-dipper-2.0-cosmos/issues).

For Hasura and BdJuno issues please report it here [https://github.com/forbole/bdjuno/issues](https://github.com/forbole/bdjuno/issues)

## License

Read our license at [https://raw.githubusercontent.com/forbole/big-dipper-2.0-cosmos/master/LICENSE](https://raw.githubusercontent.com/forbole/big-dipper-2.0-cosmos/master/LICENSE)

## Ledger and Transaction Support

While Big Dipper 2.0 no longer supports ledger or any kind of transactions in favor of [Forbole X](https://github.com/forbole/forbole-x), the original [Big Dipper](https://github.com/forbole/big-dipper) will continue have this feature.


CHANGELOG.md
# Unreleased

## Changes
- Added Sentry's Next.js SDK ([\#1005](https://github.com/forbole/big-dipper-2.0-cosmos/issues/1005))
- Switched from using `npm` to `yarn berry` package manager ([\#1028](https://github.com/forbole/big-dipper-2.0-cosmos/pull/1028))
- Added message type to transactions ([\#1034](https://github.com/forbole/big-dipper-2.0-cosmos/issues/1034))
- Incremented unit tests coverage ([\#1044](https://github.com/forbole/big-dipper-2.0-cosmos/pull/1044))


# base-v2.1.2 - 2022-09-11

## Fixes
- Fixed `/font` after chain prefix `/desmos` in global.css ([\#992](https://github.com/forbole/big-dipper-2.0-cosmos/issues/992))


# base-v2.1.1 - 2022-08-29

## Fixes
- Added missing dependency of `useEffect` in `useProfilesRecoil` to avoid desmos profile from not being loaded ([\#904](https://github.com/forbole/big-dipper-2.0-cosmos/issues/904))

## Changes
- Updated validator and account details to not be found if bech32 is invalid
- Removed the use of NEXT_PUBLIC_URL
- Updated graphql types generation structure (in preparation for third party modules)
- Updated preview image location
- Updated change url files ([\#972](https://github.com/forbole/big-dipper-2.0-cosmos/issues/972))
- Added `NEXT_PUBLIC_MATOMO_URL` and `NEXT_PUBLIC_MATOMO_SITE_ID` to github workflow production ([\#972](https://github.com/forbole/big-dipper-2.0-cosmos/issues/972))

# base-v2.1.0 - 2022-04-19

## Changes
- Updated not found and 404 logo ([\#792](https://github.com/forbole/big-dipper-2.0-cosmos/issues/792))
- Fixed params % display ([\#795](https://github.com/forbole/big-dipper-2.0-cosmos/issues/795))
- Updated params page if time period less than 1 day then show seconds ([\#797](https://github.com/forbole/big-dipper-2.0-cosmos/issues/797))
- Added token price history component ([\#784](https://github.com/forbole/big-dipper-2.0-cosmos/issues/784))
- Updated Account details hasura actions to default call 100 items instead of 10 for better performance
- Added back proposal details quorum % ([\#788](https://github.com/forbole/big-dipper-2.0-cosmos/issues/788))
- Updated validator details hasura actions performances ([\#812](https://github.com/forbole/big-dipper-2.0-cosmos/issues/812))

## Fixes
- Fixed commission being displayed for non validator accounts ([\#787](https://github.com/forbole/big-dipper-2.0-cosmos/issues/787))
- Added community tax in to apr calculation ([\#810](https://github.com/forbole/big-dipper-2.0-cosmos/issues/810))

# base-v2.0.2 - 2022-03-23

## Bug fixes
- Fixed proposal details chart percentage display

# base-v2.0.1 - 2022-03-17
## Bug fixes
- Fixed online voting power total ([\#800](https://github.com/forbole/big-dipper-2.0-cosmos/issues/800))

# base-v2.0.0 - 2022-03-10

## Changes
- Changed online voting power to be a query instead of a subscription ([\#638](https://github.com/forbole/big-dipper-2.0-cosmos/issues/638))
- Optimised rendering of Proposal Details page ([\#763](https://github.com/forbole/big-dipper-2.0-cosmos/issues/763))
- Updated online voting power display ([\#776](https://github.com/forbole/big-dipper-2.0-cosmos/issues/776))
- Improved initial loading speed by asyncing top level functions ([\#773](https://github.com/forbole/big-dipper-2.0-cosmos/issues/773))
## Bug fixes
- Fixed Apr error if bonded tokens is 0 ([\#758](https://github.com/forbole/big-dipper-2.0-cosmos/issues/758))

## Breaking
- [Bdjuno](https://github.com/forbole/bdjuno) must be on `v2.0.0`

# base-v2.0.0-rc3 - 2022-02-25

## Changes
- Changed `NEXT_PUBLIC_WS_CHAIN_URL` to `NEXT_PUBLIC_RPC_WEBSOCKET` for clarification
- Changed `NEXT_PUBLIC_CHAIN_STATUS` to `NEXT_PUBLIC_CHAIN_TYPE` for clarification

## Migration
- Changed env `NEXT_PUBLIC_WS_CHAIN_URL` to `NEXT_PUBLIC_RPC_WEBSOCKET` or don't. It's backwards compatible
- Changed env `NEXT_PUBLIC_CHAIN_STATUS` to `NEXT_PUBLIC_CHAIN_TYPE` or don't. It's backwards compatible

## Breaking
- [Bdjuno](https://github.com/forbole/bdjuno) must be on `v2.0.0`

# base-v2.0.0-rc2 - 2022-02-24

## Changes
- Updated market cap display ([\#698](https://github.com/forbole/big-dipper-2.0-cosmos/issues/698))
- Optimised validator details and account details to prevent random polling behavior ([\#703](https://github.com/forbole/big-dipper-2.0-cosmos/issues/703))
- Added hasura actions error handling in account details ([\#713](https://github.com/forbole/big-dipper-2.0-cosmos/issues/713))
- Updated to display accounts even if balance is 0 and does not exist ([\#692](https://github.com/forbole/big-dipper-2.0-cosmos/issues/692))
- Updated handling of 18 decimal places denoms ([\#724](https://github.com/forbole/big-dipper-2.0-cosmos/issues/724))

## Bug fixes
- Added better handling of get denom return types ([\#735](https://github.com/forbole/big-dipper-2.0-cosmos/issues/735))
- Updated ui to match the improved hasura actions return types

# base-v2.0.0-rc1 - 2022-02-07

## Changes
- Updated Hasura Actions

## Breaking

- [Bdjuno](https://github.com/forbole/bdjuno) must be on `v1.0.0`

# base-v1.10.0 - 2022-01-25

## Changes
- Optimized initial loading by setting basic details first then profiles after ([\#629](https://github.com/forbole/big-dipper-2.0-cosmos/issues/629))
- Added vp token unit in config ([\#645](https://github.com/forbole/big-dipper-2.0-cosmos/issues/645))

## Bug fixes
- Fixed home page validator image url not displaying correctly ([\#632](https://github.com/forbole/big-dipper-2.0-cosmos/issues/632))
- Fix validator anc account details possible infinite load due to dayjs in hook
- Added description sanitization to proposals list ([\#666](https://github.com/forbole/big-dipper-2.0-cosmos/issues/666))
- Fixed pagination hook page callback ([\#667](https://github.com/forbole/big-dipper-2.0-cosmos/issues/667))
- Showed address if name or moniker is empty ([\#668](https://github.com/forbole/big-dipper-2.0-cosmos/issues/668))

## Migration

- [v1.9.0 to v1.10.0](https://docs.bigdipper.live/cosmos-based/frontend/migrations/v1.9.0-to-v1.10.0)


jest.setup.js
/* eslint-disable */

require('@testing-library/jest-dom/extend-expect');
require('jest-localstorage-mock');

jest.mock('@utils/dayjs', () => {
  const mockTest = () => ({
    format: jest.fn(() => '2020-08-10 12:00:00'),
  });

  mockTest.utc = jest.fn(() => {
    const format = jest.fn(() => '2020-08-10 12:00:00');
    return (
      {
        format,
        fromNow: jest.fn(() => '1 day ago'),
        diff: jest.fn(() => 30),
        local: jest.fn(() => ({
          format,
        }))
      }
    );
  });

  return ({
    __esModule: true,
    default: mockTest,
    formatDayJs: jest.fn(() => '2020-08-10 12:00:00')
  });
});

jest.mock('next/dynamic', () => () => {
  const DynamicComponent = () => null;
  DynamicComponent.displayName = 'LoadableComponent';
  DynamicComponent.preload = jest.fn();
  return DynamicComponent;
});

jest.mock('@configs', () => ({
  chainConfig: {
    "title": "Desmos Block Explorer",
    "network": "morpheus-apollo-1",
    "icon": "https://raw.githubusercontent.com/forbole/big-dipper-assets/master/desmos/icon.svg?sanitize=true",
    "logo": "https://raw.githubusercontent.com/forbole/big-dipper-assets/master/desmos/logo.svg?sanitize=true",
    "prefix": {
      "consensus": "desmosvalcons",
      "validator": "desmosvaloper",
      "account": "desmos"
    },
    "genesis": {
      "time": "2021-04-27T13:00:00",
      "height": 1
    },
    "primaryTokenUnit": "udaric",
    "tokenUnits": {
      "udaric": {
        "display": "daric",
        "exponent": 6
      }
    },
    "extra": {
      "profile": true
    },
    "endpoints": {
      "graphql":"https://gql.desmos.forbole.com/v1/graphql",
      "graphqlWebsocket":"wss://gql.desmos.forbole.com/v1/graphql",
      "publicRpcWebsocket":"wss://rpc.desmos.forbole.com/websocket"
    },
    "general": {
      "basePath": "/desmos",
      "previewImage": "https://s3.bigdipper.live/desmos.png"
    },
    "marketing": {
      "matomoURL":"https://analytics.bigdipper.live",
      "matomoSiteID":"8"
    },
    "style": {
      "themes": {
        "default": "light",
        "themeList": ["light", "dark", "deuteranopia", "tritanopia"],
        "light": {
          "background": {
            "default": "#F8F8F8",
            "paper": "#FFFFFF",
            "surfaceOne": "#FFFFFF",
            "surfaceTwo": "#F8F8F8"
          },
          "primary": {
            "main": "#FD3B4C",
            "contrastText": "#fff"
          },
          "divider": "#E8E8E8",
          "text": {
            "primary": "#000000",
            "secondary": "#414141"
          },
          "fonts": {
            "fontOne": "#000000",
            "fontTwo": "#414141",
            "fontThree": "#777777",
            "fontFour": "#999999",
            "fontFive": "#FFFFFF",
            "highlight": "#1D86FF"
          },
          "primaryData": {
            "one": "#FA3A39",
            "two": "#FD5E1F",
            "three": "#FD5D4E",
            "four": "#FD9526"
          },
          "results": {
            "pass": "#1EC490",
            "fail": "#FD3B4C"
          },
          "tokenomics": {
            "one": "#1EC490",
            "two": "#497BFF",
            "three": "#9F46EC"
          },
          "condition": {
            "zero": "#E8E8E8",
            "one": "#1EC490",
            "two": "#FF961B",
            "three": "#FC6A8A"
          },
          "charts": {
            "zero": "#E8E8E8",
            "one": "#EB3AA4",
            "two": "#497BFF",
            "three": "#FF961B",
            "four": "#1EC490",
            "five": "#9F46EC"
          }
        },
        "dark": {
          "background": {
            "default": "#0A0A0A",
            "paper": "#131316",
            "surfaceOne": "#131316",
            "surfaceTwo": "#19191D"
          },
          "primary": {
            "main": "Color('#FD3B4C').alpha(0.7).string()",
            "contrastText": "#fff"
          },
          "divider": "#3D3D43",
          "text": {
            "primary": "#E6E6E6",
            "secondary": "#AAAAAB"
          },
          "fonts": {
            "fontOne": "#E6E6E6",
            "fontTwo": "#AAAAAB",
            "fontThree": "#818181",
            "fontFour": "#999999",
            "fontFive": "#FFFFFF",
            "highlight": "#1D86FF"
          },
          "primaryData": {
            "one": "#af2929",
            "two": "#b44516",
            "three": "#b14237",
            "four": "#b16919"
          },
          "results": {
            "pass": "#198a65",
            "fail": "#b12a34"
          },
          "tokenomics": {
            "one": "#1EC490",
            "two": "#497BFF",
            "three": "#9F46EC"
          },
          "condition": {
            "zero": "#E8E8E8",
            "one": "#1EC490",
            "two": "#FF961B",
            "three": "#FF608A"
          },
          "charts": {
            "zero": "#E8E8E8",
            "one": "#EB3AA4",
            "two": "#497BFF",
            "three": "#FF961B",
            "four": "#1EC490",
            "five": "#9F46EC"
          }
        }
      }
    }
  },
  generalConfig: {
    "maintainer": {
      "name": "Forbole",
      "url": "https://forbole.com"
    },
    "github": {
      "reportIssue": "https://github.com/forbole/big-dipper-2.0-cosmos/issues"
    }
  }
}));

jest.mock('@recoil/profiles', () => {
  return ({
    useProfileRecoil: jest.fn((address) => ({
      address,
      name: address,
      imageUrl: ''
    })),
    useProfilesRecoil: jest.fn((address) => ({
      address,
      name: address,
      imageUrl: ''
    })),
  });
});
